server/src/test/java/org/elasticsearch/cluster/health/ClusterHealthAllocationTests.java

```diff
@@ -0,0 +1,102 @@
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.RoutingTable;
+
+import java.util.Collections;
+
+import static org.elasticsearch.cluster.routing.ShardRoutingState.*;
+
+public class ClusterHealthAllocationTests extends ESAllocationTestCase {
+    public void testClusterHealth() {
+        ClusterState clusterState = ClusterState.builder(ClusterName.DEFAULT).build();
+        clusterState = addNode(clusterState, "node_m", true);
+        assertEquals(ClusterHealthStatus.GREEN, getClusterHealthStatus(clusterState));
+        MetaData metaData = MetaData.builder()
+            .put(IndexMetaData.builder("test")
+                .settings(settings(Version.CURRENT))
+                .numberOfShards(2)
+                .numberOfReplicas(1))
+            .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("test")).build();
+        clusterState = ClusterState.builder(clusterState).metaData(metaData).routingTable(routingTable).build();
+        MockAllocationService allocation = createAllocationService();
+        clusterState = allocation.reroute(clusterState, "reroute");
+        assertTrue(clusterState.nodes().getDataNodes().size() == 0);
+        assertTrue(clusterState.nodes().getMasterNodes().size() == 1);
+        assertEquals(ClusterHealthStatus.RED, getClusterHealthStatus(clusterState));
+        clusterState = addNode(clusterState, "node_d1", false);
+        assertTrue(clusterState.nodes().getDataNodes().size() == 1);
+        clusterState = allocation.reroute(clusterState, "reroute");
+        // starting primaries
+        clusterState = allocation.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING));
+        assertEquals(ClusterHealthStatus.YELLOW, getClusterHealthStatus(clusterState));
+        clusterState = addNode(clusterState, "node_d2", false);
+        clusterState = allocation.reroute(clusterState, "reroute");
+        // starting replicas
+        clusterState = allocation.applyStartedShards(clusterState, clusterState.getRoutingNodes().shardsWithState(INITIALIZING));
+        assertEquals(ClusterHealthStatus.GREEN, getClusterHealthStatus(clusterState));
+        clusterState = removeNode(clusterState, "node_d1");
+        assertEquals(ClusterHealthStatus.YELLOW, getClusterHealthStatus(clusterState));
+        clusterState = removeNode(clusterState, "node_d2");
+        assertEquals(ClusterHealthStatus.RED, getClusterHealthStatus(clusterState));
+        routingTable = RoutingTable.builder(routingTable).remove("test").build();
+        metaData = MetaData.builder(clusterState.metaData()).remove("test").build();
+        clusterState = ClusterState.builder(clusterState).routingTable(routingTable).metaData(metaData).build();
+        assertTrue(clusterState.nodes().getDataNodes().size() == 0);
```

Comment:
As above, prefer `assertEquals` here.

```suggestion
        assertEquals(0, clusterState.nodes().getDataNodes().size());
```